Add support for advisory collections (RhBug:1804234)

kontura opened the pull request:
This requires libsolv PR: https://github.com/openSUSE/libsolv/pull/402 (it hasn't been released yet).

https://bugzilla.redhat.com/show_bug.cgi?id=1804234

kontura committed:
Add getApplicablePackages to advisory and isApplicable to advisorymodule

Increases minimal libsolv dependency, since its needed for correct
collections handling.

kontura committed:
Keep isAdvisoryApplicable to preserve API

Its not actually used anywhere in libdnf.

kontura committed:
Properly handle multiple collections in updateinfo.xml

= changelog =
msg: Properly handle multiple collections in updateinfo.xml
type: bugfix
resolves: https://bugzilla.redhat.com/show_bug.cgi?id=1804234

kontura committed:
Add tests for Advisory, Query, DnfPackage regarding advisory collections

kontura commented:
CI tests: https://github.com/rpm-software-management/ci-dnf-stack/pull/909

j-mracek commented on data/tests/modules/etc/dnf/modules.d/base-runtime.module:
```diff
@@ -2,5 +2,6 @@
-enabled = 1
-locked = 0
\ No newline at end of file
+locked = 0
```
Please could you remove the change. It appears when dnf crashes during the test.

j-mracek commented on data/tests/modules/etc/dnf/modules.d/httpd.module:
```diff
@@ -2,5 +2,6 @@
-locked = 0
\ No newline at end of file
+locked = 0
+name=httpd
```
Please could you remove the change. It appears when dnf crashes during the test.

kontura replied:
I have removed the file changes but added a different commit that changes the ModulePackageContainerTest quite a bit in order to not change the files in repo. I can extract it into a separate PR if you prefer.

kontura committed:
Run ModulePackageContainerTest tests in tmpdir, merge interdependent

This prevents modification of:
data/tests/modules/etc/dnf/modules.d/base-runtime.module
data/tests/modules/etc/dnf/modules.d/httpd.module
which showed up in git every time the tests were run.

Some of the tests were relying on the fact that previous test modified
these files in the repo, these tests are merged or changed to work
independently.

kontura commented:
Rebased and updated the unittest to be more host independent.

j-mracek commented on libdnf/sack/advisorycollection.cpp:
```diff
@@ -0,0 +1,86 @@
+        return true;
+    }
+
+     for (auto & moduleAdvisory: pImpl->moduleList) {
```
I see 5 spaces here and not 4. Please could you check it?

kontura replied:
It was removed.

j-mracek commented on libdnf/sack/advisory.cpp:
```diff
@@ -151,9 +152,59 @@ std::vector<AdvisoryModule> Advisory::getModules() const
     return moduleList;
 }
 
+std::vector<AdvisoryCollection> Advisory::getCollections() const
```
What about to also add method, get applicable packages. From code I sow that after `getCollections()` there is an iterator that ask if collection is applicable and then it ask for collection packages.
```
                for (auto & advisoryCollection: advisory.getCollections()) {
                    if (advisoryCollection.isApplicable()) {
                        auto new_pkgs = advisoryCollection.getPackages();
```
The new approach could skip creation of collection object at all and generates only advisory packages for applicable collections. What do you think?

kontura replied:
I have used the approach we discussed in our meeting. The AdvisoryCollection class was removed. It is simpler and I think better approach.

I have also separated the unittest into separate files for AdvisoryTest, QueryTest and DnfPackageTest.

j-mracek commented on libdnf/hy-package.cpp:
```diff
@@ -1065,12 +1066,38 @@ dnf_package_get_advisories(DnfPackage *pkg, int cmp_type)
+
+            std::vector<libdnf::AdvisoryModule> moduleList;
+
+            dataiterator_seek(&di, DI_SEEK_PARENT);
```
Please could you add a comment what following three lines do? Also could you add a relation with the new dataiterator below?

kontura replied:
I have added the comment, can you check if it is detailed enough and understandable?

j-mracek replied:
Nice

j-mracek commented on libdnf/sack/advisory.cpp:
```diff
@@ -151,9 +151,66 @@ std::vector<AdvisoryModule> Advisory::getModules() const
+                isModuleCollectionApplicable = false;
+                break;
+            }
+        }
```
Somehow I believe that the logic is not the proper one. The logic says that when any module in collection is not applicable then the collection is not applicable. What about to shorten it like:
```
        bool isModuleCollectionApplicable = true;
        dataiterator_init(&di_inner, pool, 0, SOLVID_POS, UPDATE_MODULE, 0, 0);
        while (dataiterator_step(&di_inner)) {
            isModuleCollectionApplicable = false;
            dataiterator_setpos(&di_inner);
            Id name = pool_lookup_id(pool, SOLVID_POS, UPDATE_MODULE_NAME);
            Id stream = pool_lookup_id(pool, SOLVID_POS, UPDATE_MODULE_STREAM);
            Id version = pool_lookup_id(pool, SOLVID_POS, UPDATE_MODULE_VERSION);
            Id context = pool_lookup_id(pool, SOLVID_POS, UPDATE_MODULE_CONTEXT);
            Id arch = pool_lookup_id(pool, SOLVID_POS, UPDATE_MODULE_ARCH);
            AdvisoryModule moduleAdvisory(sack, advisory, name, stream, version, context, arch);
            if (moduleAdvisory.isApplicable()) {
                isModuleCollectionApplicable = true;
                break;
            }
        }
        dataiterator_free(&di_inner);
```

kontura replied:
Fixed, thank you.

kontura replied:
Ok, I have fixed the logic as well now. Only one applicable module is sufficient to make a collection applicable.

j-mracek commented on libdnf/hy-package.cpp:
```diff
@@ -1053,6 +1053,7 @@ dnf_package_get_advisories(DnfPackage *pkg, int cmp_type)
     dataiterator_init(&di, pool, 0, 0, UPDATE_COLLECTION_NAME,
```
What about to iterate over all advisories and then call advisory.getApplicablePackages(). Then search if advisory contains proper package?

j-mracek replied:
```
    dataiterator_init(&di, pool, 0, 0, 0, 0, 0);
    dataiterator_prepend_keyname(&di, UPDATE_COLLECTION);
    while (dataiterator_step(&di)) {
        dataiterator_setpos_parent(&di);
        Advisory advisory(sack, di.solvid);
        auto advisoryPkgs = advisory.getApplicablePackages()
```
Then iterate over packages and test them?

kontura replied:
I think that is a trade off between code readability and performance. It would be more understandable and simpler but a lot more objects would be created, it would make objects for all advisories and almost all their packages, this way we don't do any of that.

I have tested it with rhel appstream metadata, looping over all available packages and getting their advisories (without any modules setup). It went from ~40sec with the current implementation to ~3m30sec when constructing all the objects. (also I have noticed the `dnf_package_get_advisories` is not actually used in dnf or libdnf).

Even still I would prefer to use the faster current version.

j-mracek replied:
Good point

j-mracek commented on libdnf/hy-package.cpp:
```diff
@@ -1053,6 +1053,7 @@ dnf_package_get_advisories(DnfPackage *pkg, int cmp_type)
     dataiterator_init(&di, pool, 0, 0, UPDATE_COLLECTION_NAME,
                       pool_id2str(pool, s->name), SEARCH_STRING);
     dataiterator_prepend_keyname(&di, UPDATE_COLLECTION);
+    dataiterator_prepend_keyname(&di, UPDATE_COLLECTIONLIST);
```
Why it requires to set `UPDATE_COLLECTION` and `UPDATE_COLLECTIONLIST`?

kontura replied:
I believe it because we are searching for string `s->name` by iterating over `UPDATE_COLLECTION_NAME` and the solv structure is:
 - `UPDATE_COLLECTIONLIST`:
 - - `UPDATE_COLLECTION`:
 - - - `UPDATE_COLLECTION_NAME`: some-name

j-mracek commented on libdnf/sack/advisory.cpp:
```diff
@@ -151,9 +151,66 @@ std::vector<AdvisoryModule> Advisory::getModules() const
 }
 
+std::vector<AdvisoryPkg>
+Advisory::getApplicablePackages() const
```
Please could you consider to add optional argument `bool withFilemanes`. It is a performance optimization. See getPackages(std::vector<AdvisoryPkg> & pkglist, bool withFilemanes) above.

kontura replied:
Added, thanks. It is used from `filterAdvisory` just like with `getPackages`.

j-mracek commented on tests/libdnf/module/ModulePackageContainerTest.cpp:
```diff
@@ -4,23 +4,32 @@ CPPUNIT_TEST_SUITE_REGISTRATION(ModulePackageContainerTest);
 {
     g_autoptr(GError) error = nullptr;
+    tmpdir = g_strdup(UNITTEST_DIR);
+    mkdtemp(tmpdir);
```
There is no check of return value. When `mkdtemp` fail it will be very difficult to discover the problem.

j-mracek commented on tests/libdnf/sack/AdvisoryTest.cpp:
```diff
@@ -0,0 +1,152 @@
+    g_autoptr(GError) error = nullptr;
+
+    tmpdir = g_strdup(UNITTEST_DIR);
+    mkdtemp(tmpdir);
```
Please could you add a test that the directory was correctly created?

j-mracek commented on tests/libdnf/sack/DnfPackageTest.cpp:
```diff
@@ -0,0 +1,80 @@
+    g_autoptr(GError) error = nullptr;
+
+    tmpdir = g_strdup(UNITTEST_DIR);
+    mkdtemp(tmpdir);
```
Please could you test a return value?

j-mracek commented on tests/libdnf/sack/QueryTest.cpp:
```diff
@@ -0,0 +1,136 @@
+    g_autoptr(GError) error = nullptr;
+
+    tmpdir = g_strdup(UNITTEST_DIR);
+    mkdtemp(tmpdir);
```
Please could you test a return value?

j-mracek commented on tests/libdnf/sack/QueryTest.hpp:
```diff
@@ -0,0 +1,32 @@
+private:
+    DnfSack *sack;
+    HyRepo repo;
+    char* tmpdir;
```
I believe that valgrind will complain that variables ware not correctly initialized. Please could you set for all values a default value nullptr?

j-mracek commented on tests/libdnf/sack/DnfPackageTest.hpp:
```diff
@@ -0,0 +1,30 @@
+private:
+    DnfSack *sack;
+    HyRepo repo;
+    char* tmpdir;
```
I believe that valgrind will complain that variables ware not correctly initialized. Please could you set for all values a default value nullptr?

j-mracek commented on tests/libdnf/sack/AdvisoryTest.hpp:
```diff
@@ -0,0 +1,57 @@
+    DnfSack *sack;
+    HyRepo repo;
+    libdnf::Advisory *advisory;
+    char* tmpdir;
```
I believe that valgrind will complain that variables ware not correctly initialized. Please could you set for all values a default value nullptr?

kontura commented:
Added `nullptr` initialization and asserting of return values from `mkdtemp`. :+1:

j-mracek commented:
LGTM

j-mracek commented:
bors try

j-mracek commented:
It looks like that the patch is not compatible with dnf unit tests. Please could you take a look? See https://download.copr.fedorainfracloud.org/results/rpmsoftwaremanagement/rpm-gitoverlay-1613576172.706418/fedora-32-x86_64/01982092-dnf/builder-live.log.gz.

kontura commented:
The problem should be fixed now. The new function `getApplicablePackages` was returning a vector and overriding the old one on each call (in `filterAdvisory` and `getAdvisoryPkgs`) instead the pkgs should be added. I have changed `getApplicablePackages` to take a vector as a paramater and add pkgs to it instead, just like original `getPackages`.

bors try
